plutus-core/plutus-core/src/PlutusCore/ParserCommon.hs

```diff
@@ -0,0 +1,202 @@
+
+whitespace :: Parser ()
+whitespace = Lex.space space1 (Lex.skipLineComment "--") (Lex.skipBlockCommentNested "{-" "-}")
+
+lexeme :: Parser a -> Parser a
+lexeme = Lex.lexeme whitespace
+
+symbol :: T.Text -> Parser T.Text
+symbol = Lex.symbol whitespace
+
+lparen :: Parser T.Text
+lparen = symbol "("
+rparen :: Parser T.Text
+rparen = symbol ")"
+
+lbracket :: Parser T.Text
+lbracket = symbol "["
+rbracket :: Parser T.Text
+rbracket = symbol "]"
+
+lbrace :: Parser T.Text
+lbrace = symbol "{"
+rbrace :: Parser T.Text
+rbrace = symbol "}"
+
+inParens :: Parser a -> Parser a
+inParens = between lparen rparen
+
+inBrackets :: Parser a -> Parser a
+inBrackets = between lbracket rbracket
+
+inBraces :: Parser a -> Parser a
+inBraces = between lbrace rbrace
+
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c = isAlphaNum c || c == '_' || c == '\''
+
+-- | Create a parser that matches the input word and returns its source position.
+-- This is for attaching source positions to parsed terms/programs.
+-- getSourcePos is not cheap, don't call it on matching of every token.
+reservedWord ::
+    -- | The word to match
+    T.Text -> Parser SourcePos
+reservedWord w = lexeme $ try $ do
+    p <- getSourcePos
+    void $ string w
+    notFollowedBy (satisfy isIdentifierChar)
+    return p
+
+builtinFunction :: (Bounded fun, Enum fun, Pretty fun) => Parser fun
+builtinFunction = lexeme $ choice $ map parseBuiltin [minBound .. maxBound]
+    where parseBuiltin builtin = try $ string (display builtin) >> pure builtin
+
+version :: Parser (PLC.Version SourcePos)
+version = lexeme $ do
+    p <- getSourcePos
+    x <- Lex.decimal
+    void $ char '.'
+    y <- Lex.decimal
+    void $ char '.'
+    PLC.Version p x y <$> Lex.decimal
+
+name :: Parser PLC.Name
+name = lexeme $ try $ do
+    void $ lookAhead letterChar
+    str <- takeWhileP (Just "identifier") isIdentifierChar
+    PLC.Name str <$> intern str
+
+tyName :: Parser PLC.TyName
+tyName = PLC.TyName <$> name
+
+-- | Turn a parser that can succeed without consuming any input into one that fails in this case.
+enforce :: Parser a -> Parser a
+enforce p = do
+    (input, x) <- match p
+    guard . not $ T.null input
+    pure x
+
+-- | Consume a chunk of text and either stop on whitespace (and consume it) or on a \')\' (without
+-- consuming it). We need this for collecting a @Text@ to pass it to 'PLC.parse' in order to
+-- parse a built-in type or a constant. This is neither efficient (because of @manyTill anySingle@),
+-- nor future-proof (what if some future built-in type has parens in its syntax?). A good way of
+-- resolving this would be to turn 'PLC.parse' into a proper parser rather than just a function
+-- from @Text@, but PLC's parser also uses 'PLC.parse' and it's currently not @megaparsec@-based,
+-- and so mixing two distinct parsing machines doesn't sound too exciting.
+--
+-- Note that this also fails on @(con string \"yes (no)\")@ as well as @con unit ()@, so it really
```

Comment:
We've got plenty of parentheses around already, so I'd be perfectly happy to throw away `()` and replace it with `unitval` or `one` or `*` or something if it would make things simpler.  Unfortunately I think we really do need to cope with parentheses inside literal strings and characters though.

Reply:
Added to comments.